doc: add instruction of Open vSwitch (#3143)
projectacrn/acrn-hypervisor: doc/tutorials/open_vswitch.rst

- Fine-grained QoS control
- Support for HFSC qdisc
- Per VM interface traffic policing
- NIC bonding with source-MAC load balancing, active backup, and L4 hashing
- OpenFlow protocol support (including many extensions for virtualization)
- IPv6 support
- Multiple tunneling protocols (GRE, VXLAN, STT, and Geneve, with IPsec support)
- Remote configuration protocol with C and Python bindings
- Kernel and user-space forwarding engine options
- Multi-table forwarding pipeline with flow-caching engine
- Forwarding layer abstraction to ease porting to new software and hardware platforms

For more detailed OVS information, please refer to `<https://www.openvswitch.org/>`_

.. _enable_ovs_in_ACRN:

How to enable OVS in ACRN
*************************
The OVS service is integrated in the ClearLinux ``service-os`` bundle.
After booting the ACRN Service OS, start the OVS service with the command::

    systemctl start openvswitch

The default ``acrn-br0`` bridge is created by the SOS ``systemd`` and
supports the UOS network.

.. figure:: images/default-acrn-network.png
   :align: center

   Default ACRN Network

.. note::
   If ``systemctl start openvswitch`` fails to start OVS service,

>    you may also need to add ``network-basic`` bundle for other network dependencies.

Reviewer: ```suggestion
   you may also need to add ``network-basic`` bundle for other network dependencies
   and run the ``systemctl`` command again.
```

How to use OVS bridge instead of ACRN bridge
********************************************
#. Disable acrn network configuration::

   # cd /usr/lib/systemd/network/
   # mv 50-acrn.network 50-acrn.network_bak

#. Modify ``50-eth.network`` to enable DHCP on OVS bridge

   .. code-block:: none

      [Match]
      Name=ovs-br0

      [Network]
      DHCP=ipv4

#. Create OVS bridge::

   # ovs-vsctl add-br ovs-br0

#. Add ``ens4``, ``acrn_tap_LaaG`` into OVS bridge::

   # ovs-vsctl add-port ovs-br0 ens4
   # ovs-vsctl add-port ovs-br0 acrn_tap_LaaG

> #. Modify ``launch_uos.sh`` script to enable ``acrn_tap_LaaG`` device before launch UOS:

Reviewer: ```suggestion
#. Modify ``launch_uos.sh`` script to enable ``acrn_tap_LaaG`` device before launching the UOS:
```

   .. code-block:: none

      # sed -i "s/virtio-net,tap0/virtio-net,acrn_tap_LaaG/" /usr/share/acrn/samples/nuc/launch_uos.sh

   .. note::
      If you set up the UOS via `acrn_quick_setup.sh
      <https://raw.githubusercontent.com/projectacrn/acrn-hypervisor/master/doc/getting-started/acrn_quick_setup.sh>`_,

>       then replace `/usr/share/acrn/samples/nuc/launch_uos.sh` with `/root/launch_uos_<version>.sh`

Reviewer: ```suggestion
      then replace ``/usr/share/acrn/samples/nuc/launch_uos.sh`` with ``/root/launch_uos_<version>.sh``
```

>       in above `sed` command.

Reviewer: ```suggestion
      in ``sed`` command above.
```

> #. The UOS and SOS network can work after reboot the host via ``ovs-br0``

Reviewer: ```suggestion
#. The UOS and SOS network will work after rebooting the host via ``ovs-br0``
```

.. _open_vswitch:

Using OVS bridge instead of ACRN bridge
#######################################
Hypervisors need the ability to bridge network traffic between VMs
and with the outside world. This tutorial describes how to

> use Open Virtual Switch (OVS) bridge instead of ACRN

Reviewer 2: Can you add a link to their website: https://www.openvswitch.org/

bridge for this purpose.

.. note::

>    OVS is integrated in the ClearLinux ``service-os``

Reviewer 2: Rather than "integrated", I would say it is provided as part of the `service-os` bundle (but it's not active/configured by default).

   bundle.  Use ClearLinux OS version ``29660`` or higher.

What is OVS
***********
Open vSwitch (OVS) is an open-source implementation of
a distributed virtual multilayer switch that provides a switching
stack for hardware virtualization environments. OVS supports multiple
protocols and standards used in computer networks. For more detailed
OVS information, please refer to `what-is-ovs
<http://docs.openvswitch.org/en/latest/intro/what-is-ovs/#what-is-open-vswitch>`_.

Why OVS
*******
Open vSwitch is targeted at multi-server virtualization deployments,

> a landscape for the built-in L2 switch (the `Linux bridge

Reviewer: ```suggestion
a landscape not well suited for ACRN's built-in L2 switch (the `Linux bridge
```

> <https://wiki.linuxfoundation.org/networking/bridge>`_) stack is not well suited.

Reviewer: ```suggestion
<https://wiki.linuxfoundation.org/networking/bridge>`_).
```

- BFD and 802.1ag link monitoring
- STP (IEEE 802.1D-1998) and RSTP (IEEE 802.1D-2004)
- Fine-grained QoS control
- Support for HFSC qdisc
- Per VM interface traffic policing
- NIC bonding with source-MAC load balancing, active backup, and L4 hashing
- OpenFlow protocol support (including many extensions for virtualization)
- IPv6 support
- Multiple tunneling protocols (GRE, VXLAN, STT, and Geneve, with IPsec support)
- Remote configuration protocol with C and Python bindings
- Kernel and user-space forwarding engine options
- Multi-table forwarding pipeline with flow-caching engine
- Forwarding layer abstraction to ease porting to new software and hardware platforms

For more detailed OVS information, please refer to `<https://www.openvswitch.org/>`_

.. _enable_ovs_in_ACRN:

How to enable OVS in ACRN
*************************

> The OVS service is integrated in the ClearLinux ``service-os`` bundle.

Reviewer 2: I would recommend we say it's installed with the `service-os` bundle instead. @dbkinder would this work for you? (the subtlety here is that it's installed but not active/enable).

Reviewer: ok.  saying "installed" instead of "integrated" works for me.

> After booting the ACRN Service OS, start the OVS service with the command::

Reviewer 2: Any update to the system will break this set-up. The GSG has instructions to disable swupd autoupdate and the GSG script does that too. *But* running the GSG script to update the SOS will break that set-up (since it will restore the default bridge). We should add a note to that effect here.

Author: Thanks for mentioning it, I will add an note in next release.

>     systemctl start openvswitch

Reviewer 2: You indicate the prompt in all other examples below so let's use it here too.

```suggestion
    # systemctl start openvswitch
```

> The default ``acrn-br0`` bridge is created by the SOS ``systemd`` and

Reviewer 2: I believe we're moving away from the term SOS (Service OS) and I think it's preferable to use "Service VM".

supports the UOS network.

.. figure:: images/default-acrn-network.png
   :align: center

   Default ACRN Network

.. note::

>    If ``systemctl start openvswitch`` fails to start OVS service,

Reviewer 2: Should we mention that if people want this by default, they should also use `systemctl enable` so OVS is started automatically after a reboot?

Author: Sure, I will update it in next release

>    you may also need to add ``network-basic`` bundle for other network dependencies

Reviewer 2: Can we determine if it's needed or not? It sounds odd that we *may* (or *may not*?) need this. Is there any case you wouldn't need this bundle to be installed?

Author: ``network-basic`` is not needed in this instruction, so I will remove this note from next release.

   and run the ``systemctl`` command again.

How to use OVS bridge instead of ACRN bridge
********************************************
#. Disable acrn network configuration::

   # cd /usr/lib/systemd/network/
   # mv 50-acrn.network 50-acrn.network_bak

#. Modify ``50-eth.network`` to enable DHCP on OVS bridge

   .. code-block:: none

      [Match]
      Name=ovs-br0

      [Network]
      DHCP=ipv4

#. Create OVS bridge::

   # ovs-vsctl add-br ovs-br0

#. Add ``ens4``, ``acrn_tap_LaaG`` into OVS bridge::

   # ovs-vsctl add-port ovs-br0 ens4
   # ovs-vsctl add-port ovs-br0 acrn_tap_LaaG

#. Modify ``launch_uos.sh`` script to enable ``acrn_tap_LaaG`` device before launching the UOS:

   .. code-block:: none

      # sed -i "s/virtio-net,tap0/virtio-net,acrn_tap_LaaG/" /usr/share/acrn/samples/nuc/launch_uos.sh

   .. note::
      If you set up the UOS via `acrn_quick_setup.sh
      <https://raw.githubusercontent.com/projectacrn/acrn-hypervisor/master/doc/getting-started/acrn_quick_setup.sh>`_,
      then replace ``/usr/share/acrn/samples/nuc/launch_uos.sh`` with ``/root/launch_uos_<version>.sh``
      in ``sed`` command above.

> #. The UOS and SOS network will work after rebooting the host via ``ovs-br0``

Reviewer 2: Update this to use "User VM" and "Service VM" instead of "UOS" and "SOS" (throughout this doc).

- IETF Auto-Attach SPBM and rudimentary required LLDP support
- BFD and 802.1ag link monitoring
- STP (IEEE 802.1D-1998) and RSTP (IEEE 802.1D-2004)
- Fine-grained QoS control
- Support for HFSC qdisc
- Per VM interface traffic policing
- NIC bonding with source-MAC load balancing, active backup, and L4 hashing
- OpenFlow protocol support (including many extensions for virtualization)
- IPv6 support
- Multiple tunneling protocols (GRE, VXLAN, STT, and Geneve, with IPsec support)
- Remote configuration protocol with C and Python bindings
- Kernel and user-space forwarding engine options
- Multi-table forwarding pipeline with flow-caching engine
- Forwarding layer abstraction to ease porting to new software and hardware platforms

For more detailed OVS information, please refer to `<https://www.openvswitch.org/>`_

.. _enable_ovs_in_ACRN:

How to enable OVS in ACRN
*************************
The OVS service is installed with the ``service-os`` bundle.
After booting the ACRN Service OS, start the OVS service with the command::

   # systemctl start openvswitch

To start OVS automatically after a reboot, you should also use this command::

   # systemctl enable openvswitch

> .. note::

Reviewer: This note should be moved to the beginning of this section, and not be a "note":

```
How to enable OVS in ACRN
*************************
The OVS service is included with the Clear Linux ``service-os`` bundle.

After booting the ACRN Service OS, disable the Clear Linux
autoupdate feature before setting up the OVS bridge to
prevent autoupdate from restoring the default bridge after
a system update::

   # swupd autoupdate --disable

You can then start the OVS service with the command::

   # systemctl start openvswitch

To start OVS automatically after a reboot, you should also use this command::

   # systemctl enable openvswitch
```

   We recommend use ``swupd autoupdate --disable`` to

>    disable ``autoupdate`` feature before set up OVS bridge

Reviewer: ```suggestion
   disable ``autoupdate`` feature before setting up OVS bridge
```

>    since it will restore the default bridge after system upgrade.

Reviewer: ```suggestion
   since autoupdate will restore the default bridge after system upgrade.
```

.. _open_vswitch:

Using OVS bridge instead of ACRN bridge
#######################################
This tutorial describes how to use Open Virtual Switch

> (OVS) bridge instead of ACRN bridge.

Reviewer 3: OVS can support 802.1Q feature to setup vlan network, and ACRN bridge can not meet this requirement.

And OVS can manage the network more easily than Linux bridge(ACRN bridge), customer wants to use it . So ACRN bridge still acts as the default network switch, and customer can replace it with OVS by requirement.

.. note::
   The OVS is integrated in ClearLinux ``service-os``
   bundle, and the ClearLinux OS must be at least ``28900`` version.

What is OVS
***********

> Open vSwitch, abbreviated as OVS, is an open-source implementation of

Reviewer: @lirui34 I see many comments are marked as resolved, but I don't see changes to the document?  Maybe an updating push wasn't done?

> Open vSwitch, abbreviated as OVS, is an open-source implementation of

Author: Oh, sorry about that, I just requested an internal review in our group yesterday so that they could only see those 'unresolved' comments.
I have pushed an new version just now :smile:

a distributed virtual multilayer switch which is used to provide a switching
stack for hardware virtualization environments, while supporting multiple
protocols and standards used in computer networks.

Open vSwitch supports the following features:

> - Visibility into inter-VM communication via NetFlow,

Reviewer 3: Yes, ACRN bridge does support the features .
The OVS does not replace ACRN bridge, we only integrate OVS in service-os bundle, and a guide for OVS setup, the users of OVS should be clear about the usage scenarios of OVS and enable OVS according to our guide.

> - Visibility into inter-VM communication via NetFlow,

Reviewer: You say "OVS does not replace ACRN bridge", but the title of this article is "Using OVS bridge **instead of** ACRN bridge", and your comment in the intro says "customer can replace it with OVS by requirement"?

When I read this section, it sounds like OVS is better than ACRN bridge because OVS supports all these listed features that aren't supported by ACRN bridge.  Why are you listing all these features if ACRN bridge supports them too?  You do mention that "OVS can support 802.1Q feature to setup vlan network, and ACRN bridge can not meet this requirement." Are there other differences?

We should add something to the introduction (lines 5-8) to explain why someone might want to replace ACRN bridge with OVS.

> - Visibility into inter-VM communication via NetFlow,

Reviewer 3: Agree with your point,  I want to change the title from "Using OVS bridge instead of ACRN bridge" to "How to enable OVS in ACRN" and remove all features introduction of OVS, since I don't want to indicate that OVS is better than ACRN brige, I just only want to introduce how to enable OVS in ACRN, and customer should know the difference between them.

what about your opinion?

> - Visibility into inter-VM communication via NetFlow,

Reviewer 2: I think this is the right tone for this document, i.e.: explain how someone can use ACRN with OVS.

Is there any such thing as an "ACRN bridge" by the way? My (possibly limited) understanding was that these are all standard Linux mechanisms that we leverage and set-up correctly in our ACRN scripts and instructions. If true, I think a good intro would be to explain that by default, we provide scripts and configuration files that set-up such bridge (https://wiki.linuxfoundation.org/networking/bridge) and that this document explains how to use OVS instead (and "instead" is the right word here, it's one or the other).

I also believe it's useful to have a few top-level reasons mentionned in the intro as to why someone may want to use OVS. The point being that if the Linux bridge is enough, there is no reason to change it but if more advanced features are required then OVS may be a better solution for them. This page on the OVS website is dedicated to explaining why and when using OVS would make sense. I would recommend we refer to that document as well in our introduction.

> - Visibility into inter-VM communication via NetFlow,

Reviewer: I agree with both of you.  Remove all the details about OVS but keep a simple overview. Changing the title makes sense too. @liuyuan-GitHub can you incorporate @gvancuts comments in a new version of this introduction.

- STP (IEEE 802.1D-1998) and RSTP (IEEE 802.1D-2004)
- Fine-grained QoS control
- Support for HFSC qdisc
- Per VM interface traffic policing
- NIC bonding with source-MAC load balancing, active backup, and L4 hashing
- OpenFlow protocol support (including many extensions for virtualization)
- IPv6 support
- Multiple tunneling protocols (GRE, VXLAN, STT, and Geneve, with IPsec support)
- Remote configuration protocol with C and Python bindings
- Kernel and user-space forwarding engine options
- Multi-table forwarding pipeline with flow-caching engine
- Forwarding layer abstraction to ease porting to new software and hardware platforms

For more detailed OVS information, please refer to `<https://www.openvswitch.org/>`_

.. _enable_ovs_in_ACRN:

How to enable OVS in ACRN
*************************
The OVS service is integrated in ClearLinux ``service-os`` bundle.
After boot ACRN Service OS, run ``systemctl start openvswitch`` command to start OVS.
And ``acrn-br0`` is a default bridge, which is created by ``systemd`` of SOS,
it is used to support UOS network.

.. figure:: images/default-acrn-network.png
   :align: center

   Default ACRN Network

.. note::

>    If ``systemctl start openvswitch`` fails to start OVS service,

Author: Rewrite this note to:
'''
If ``systemctl start openvswitch`` fails to start OVS service,
"you may also need to add ``network-basic`` bundle for other network dependencies.
'''

   you may need to install ``openvswitch`` via ``swupd bundle-add network-basic`` command.

> How to use OVS bridge instead ACRN bridge

Reviewer 3: This does not need, since OVS only control its created bridge.

*****************************************
#. Disable acrn network configuration::

   # cd /usr/lib/systemd/network/

>    # mv 50-acrn.network 50-acrn.network_bak

Author: It should be ``mv`` command because we just need to edit ``50-eth.network`` but not ``50-acrn.network``

>    # mv 50-acrn.network 50-acrn.network_bak

Reviewer: Oh, I missed the file name was different.  Nevermind :)

#. Modify ``50-eth.network`` to enable DHCP on OVS bridge

    .. code-block:: none

       [Match]
       Name=ovs-br0

       [Network]
       DHCP=ipv4

#. Create OVS bridge::

   # ovs-vsctl add-br ovs-br0

#. Add ``ens4``, ``acrn_tap_LaaG`` into OVS bridge::

   # ovs-vsctl add-port ovs-br0 ens4
   # ovs-vsctl add-port ovs-br0 acrn_tap_LaaG

> #. Modify ``launch_uos.sh`` script to enable ``acrn_tap_LaaG`` device before launch UOS::

Author: Quick setup script still use ``tap0`` as the default ``virtio-net`` device name, so the ``sed`` command is also work for the ``launch_uos_<version>.sh`` script.

> #. Modify ``launch_uos.sh`` script to enable ``acrn_tap_LaaG`` device before launch UOS::

Reviewer: OK, thanks.  Do we need to mention that the user might need to update the ``launch_uos_<version>.sh`` script here?

> #. Modify ``launch_uos.sh`` script to enable ``acrn_tap_LaaG`` device before launch UOS::

Author: Yes, I will add a ``note`` to inform user to replace launch uos script directory in the ``sed`` command in case they're using the quick setup script to setup the UOS.

   # sed -i "s/virtio-net,tap0/virtio-net,acrn_tap_LaaG/" /usr/share/acrn/samples/nuc/launch_uos.sh

#. The UOS and SOS network can work after reboot the host via ``ovs-br0``

Example for OVS in ACRN
***********************
We will utilize VLAN feature by OVS bridge to support network isolation
between UOSs. Below Figure shows an example, there are four UOSs in two hosts,
hosts are directly connected by a network cable, UOSs are interconnected
through statically configured IP addresses, and use VLAN id to put UOS_1 of
HOST_1 and UOS_1 of HOST_2 into a VLAN. Similarly, put UOS_2 of HOST_1 and UOS_2 of
HOST_2 into a VLAN, then UOS_1s can communicate with each other,
and UOS_2s can directly communicate with each other, but UOS_1s and UOS_2s cannot connect.

.. figure:: images/example-of-OVS-usage.png
   :align: center

   An example of OVS usage in ACRN

> You can follow these steps to setup OVS network on both HOSTs:

Author: I will remove the duplicate lines in next release.

#. Install network tools::

   # swupd bundle-add network-basic

#. Setup ``ovs-br0`` instead of ``acrn-br0``, you can refer to the :ref:`enable_ovs_in_ACRN` for details.

   a. Start openvswitch::

      # systemctl start openvswitch

   #. Disable acrn network configuration::

      # cd /usr/lib/systemd/network/
      # mv 50-acrn.network 50-acrn.network_bak

   #. Modify ``50-eth.network`` to enable DHCP on OVS bridge:

      .. code-block:: none

         [Match]
         Name=ovs-br0

         [Network]
         DHCP=ipv4

   #. Create OVS bridge::

      # ovs-vsctl add-br ovs-br0

   #. Add ``ens4``, ``acrn_tap_LaaG<UOS number>`` into OVS bridge::

      # ovs-vsctl add-port ovs-br0 ens4
      # ovs-vsctl add-port ovs-br0 acrn_tap_LaaG1 tag=101
      # sed -i "s/virtio-net,tap0/virtio-net,acrn_tap_LaaG1/" /usr/share/acrn/samples/nuc/launch_uos.sh
      # reboot

> #. Configure the static IP address on both HOSTs and it's UOSs::

Author: Yes, I will remove the duplicate lines in next release.
